dipy/tracking/local/localtrack.pyx

```diff
@@ -208,3 +199,301 @@ cdef int _local_tracker(
         # maximum length of streamline has been reached, return everything
         i = streamline.shape[0]
     return i
+
+
+def pft_tracker(
+        DirectionGetter dg,
+        ConstrainedTissueClassifier tc,
+        np.float_t[:] seed_pos,
+        np.float_t[:] first_step,
+        np.float_t[:] voxel_size,
+        np.float_t[:, :] streamline,
+        np.float_t[:, :] directions,
+        double step_size,
+        int pft_nbr_back_steps,
+        int pft_max_steps,
+        int pft_max_trials,
+        int particle_count,
+        np.float_t[:, :, :, :] particle_paths,
+        np.float_t[:, :, :, :] particle_dirs,
+        np.float_t[:, :] particle_weights,
+        np.int_t[:, :, :]  particle_states):
+    """Tracks one direction from a seed using the particle filtering algorithm.
+
+    This function is the main workhorse of the ``ParticleFilteringTracking``
+    class defined in ``dipy.tracking.local.localtracking``.
+
+    Parameters
+    ----------
+    dg : DirectionGetter
+        Used to choosing tracking directions.
+    tc : TissueClassifier
+        Used to check tissue type along path.
+    seed_pos : array, float, 1d, (3,)
+        First point of the (partial) streamline.
+    first_step : array, float, 1d, (3,)
+        Initial seeding direction. Used as ``prev_dir`` for selecting the step
+        direction from the seed point.
+    voxel_size : array, float, 1d, (3,)
+        Size of voxels in the data set.
+    streamline : array, float, 2d, (N, 3)
+        Output of tracking will be put into this array. The length of this
+        array, ``N``, will set the maximum allowable length of the streamline.
+    directions : array, float, 2d, (N, 3)
+        Output of tracking directions will be put into this array. The length
+        of this array, ``N``, will set the maximum allowable length of the
+        streamline.
+    step_size : float
+        Size of tracking steps in mm if ``fixed_step``.
+    pft_nbr_back_steps : int
+        Number of tracking steps to back track before starting the particle
+        filtering tractography.
+    pft_max_steps : int
+        Maximum number of steps for the particle filtering tractography.
+    pft_max_trials : int
+        Maximum number of trials for the particle filtering tractography
+        (Prevents infinite loops).
+    particle_count : int
+        Number of particles to use in the particle filter.
+    particle_paths : array, float, 4d, (2, particle_count, pft_max_steps, 3)
+        Temporary array for paths followed by all particles.
+    particle_dirs : array, float, 4d, (2, particle_count, pft_max_steps, 3)
+        Temporary array for directions followed by particles.
+    particle_weights : array, float, 2d, (2, particle_count)
```

Comment:
I don't understand why you need the starting `(2` here. I see that you used it in the Resampling part, but all weights are copied from 0 to 1, cumsum+other in 1, then you overwrite 0 with new data. Why couldn't you work in 0 directly?

It seems to be required for the other temporary arrays though so I won't complain :)